src/main/windows/callsWidgetWindow.ts

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {EventEmitter} from 'events';
+import {BrowserWindow, Rectangle, ipcMain, IpcMainEvent} from 'electron';
+import log from 'electron-log';
+
+import {
+    CallsWidgetWindowConfig,
+    CallsWidgetResizeMessage,
+    CallsWidgetShareScreenMessage,
+} from 'types/calls';
+
+import {getLocalPreload} from 'main/utils';
+
+import {PRODUCTION} from 'common/utils/constants';
+import Utils from 'common/utils/util';
+import {
+    CALLS_WIDGET_RESIZE,
+    CALLS_WIDGET_SHARE_SCREEN,
+} from 'common/communication';
+
+type LoadURLOpts = {
+    extraHeaders: string;
+}
+
+function boundsDiff(base: Rectangle, actual: Rectangle) {
+    return {
+        x: base.x - actual.x,
+        y: base.y - actual.y,
+        width: base.width - actual.width,
+        height: base.height - actual.height,
+    };
+}
+
+export default class CallsWidgetWindow extends EventEmitter {
+    public win: BrowserWindow;
+    private main: BrowserWindow;
+    private config: CallsWidgetWindowConfig;
+    private minWidth = 280;
+    private minHeight = 86;
+    private pluginID = 'com.mattermost.calls';
+    private boundsErr: Rectangle = {
+        x: 0,
+        y: 0,
+        width: 0,
+        height: 0,
+    };
+    private offsetsMap = {
+        'calls-widget-menu': {
+            height: 0,
+        },
+    };
+
+    constructor(mainWindow: BrowserWindow, config: CallsWidgetWindowConfig) {
+        super();
+
+        this.config = config;
+        this.main = mainWindow;
+        this.win = new BrowserWindow({
+            width: this.minWidth,
+            height: this.minHeight,
+            minWidth: this.minWidth,
+            minHeight: this.minHeight,
+            title: 'Calls Widget',
+            fullscreen: false,
+            resizable: false,
+            frame: false,
+            transparent: true,
+            show: false,
+            alwaysOnTop: true,
+            webPreferences: {
+                preload: getLocalPreload('callsWidget.js'),
+            },
+        });
+
+        this.win.once('ready-to-show', () => this.win.show());
+        this.win.once('show', this.onShow);
+        this.win.on('closed', this.onClosed);
+        ipcMain.on(CALLS_WIDGET_RESIZE, this.onResize);
+        ipcMain.on(CALLS_WIDGET_SHARE_SCREEN, this.onShareScreen);
+
+        this.load();
+    }
+
+    public close() {
+        log.debug('CallsWidgetWindow.close');
+        this.win.close();
+    }
+
+    public getServerName() {
+        return this.config.serverName;
+    }
+
+    private load() {
+        const opts = {} as LoadURLOpts;
+        if (Utils.runMode() !== PRODUCTION) {
```

Comment:
Sure. I am actually still figuring out if we need it in the first place. The problem I was facing was that the window would load stale (client cached) data and I had to hard refresh every time so using that header did that automatically.

Reply:
Okay so the issue turned out to be even worse with a proxy in between as it would cache stuff almost permanently even when updating the plugin since the `widget.html` file was considered static (and served as such). I've now changed the way we are serving the widget app entirely, instead of relying on the static folder I'll be using a custom plugin API endpoint `/widget` to do it so that we get more control on how assets are served and also the caching now works as expected, only updating when the files are modified.